File: examples/02_borrowing/main.cpp
// ============================================
// 02 — Borrowing & References (C++)
// ============================================
// Kompilieren: g++ -std=c++17 main.cpp -o main && ./main

#include <iostream>
#include <string>
#include <vector>

// --- Beispiel 1: const Referenz (immutable borrow) ---
size_t calculate_length(const std::string& s) {
    return s.length(); // s wird nicht verändert, nicht kopiert
}

// --- Beispiel 2: Referenz (mutable borrow) ---
void change(std::string& s) {
    s += ", world";
}

// --- Beispiel 3: Das Problem — C++ verhindert NICHTS ---
void aliasing_problem() {
    std::string s = "hello";

    std::string& r1 = s; // mutable Referenz
    std::string& r2 = s; // zweite mutable Referenz — erlaubt in C++!

    r1 += " from r1";
    r2 += " from r2"; // gleichzeitiger Zugriff — kein Fehler

    std::cout << s << std::endl;
    // In Rust: ❌ COMPILE ERROR — nur eine &mut Referenz erlaubt
}

// --- Beispiel 4: Dangling Reference — Compiler verhindert es nicht ---
const std::string& get_dangling() {
    std::string local = "ich werde zerstört";
    return local; // ⚠️ Dangling Reference — UB!
    // Compiler warnt, verhindert aber nicht
}

// --- Beispiel 5: Iterator Invalidation ---
void iterator_invalidation() {
    std::vector<int> v = {1, 2, 3};
    auto& first = v[0]; // Referenz auf erstes Element

    v.push_back(4); // Reallokation! first ist jetzt DANGLING
    // std::cout << first; // 💥 Undefined Behavior
    // In Rust: ❌ COMPILE ERROR — kann nicht passieren
}

int main() {
    std::string s = "hello";

    std::cout << "Länge: " << calculate_length(s) << std::endl;
    std::cout << "Vorher: " << s << std::endl;

    change(s);
    std::cout << "Nachher: " << s << std::endl;

    std::cout << "\n=== Aliasing (in Rust verboten) ===" << std::endl;
    aliasing_problem();

    return 0;
}

File: examples/02_borrowing/main.rs
// ============================================
// 02 — Borrowing & Referenzen
// ============================================
//
// Borrowing = Ausleihen ohne Ownership zu übertragen.
// Der Owner bleibt Eigentümer — die Referenz ist nur temporär.
//
// Regeln (Compile-Zeit geprüft, kein Runtime-Overhead):
//   &T     → immutable borrow  — beliebig viele gleichzeitig
//   &mut T → mutable borrow    — genau eine, exklusiv

fn main() {
    // --------------------------------------------------
    // 1. Immutable Borrow — Owner bleibt Eigentümer
    // --------------------------------------------------
    let s = String::from("hello");

    let laenge = berechne_laenge(&s); // & = borrow, kein Move
    println!("'{}' hat {} Zeichen", s, laenge); // ✅ s gehört noch main

    // --------------------------------------------------
    // 2. Mehrere Leser gleichzeitig — kein Problem
    // --------------------------------------------------
    let r1 = &s;
    let r2 = &s;
    let r3 = &s;
    println!("{} {} {}", r1, r2, r3); // ✅ alle lesen gleichzeitig

    // --------------------------------------------------
    // 3. Mutable Borrow — exklusiver Schreibzugriff
    // --------------------------------------------------
    let mut text = String::from("hello");

    aendern(&mut text);
    println!("Geändert: {}", text); // ✅

    // --------------------------------------------------
    // 4. Nicht gleichzeitig lesen & schreiben
    // --------------------------------------------------
    let mut data = String::from("hello");

    let leser = &data; // ✅ immutable borrow
    // aendern(&mut data);      // ❌ COMPILE ERROR: cannot borrow as mutable
                                //    because it is also borrowed as immutable
    println!("{}", leser); // leser wird hier nicht mehr genutzt

    aendern(&mut data); // ✅ jetzt OK — leser ist nicht mehr aktiv
    println!("{}", data);
}

fn berechne_laenge(s: &String) -> usize {
    s.len() // s wird nicht gedroppt — nur geliehen
}

fn aendern(s: &mut String) {
    s.push_str(", world");
}

// Diese Funktion kompiliert NICHT — Dangling Reference:
// fn dangle() -> &String {
//     let s = String::from("hello");
//     &s   // ❌ s wird am Ende des Scopes gedroppt
// }        //    die Referenz würde auf ungültigen Speicher zeigen
